Separated login and login_info routes

marcus-oscarsson opened the pull request:
Separated `login` and `login_info` routes. Returning `401 - Unauthorized` from login_info instead of redirect. Removed the call to `requireAuth` from the `PrivateOutlet` component as it runs during rendering. 

`requireAuth` is now executed on page load (`useEffect` of `App`) and on change of route using the loader functionality of `react-router`. The later makes the transition from login to application much smoother, as no reload of the page is needed. 

The proposal selection dialog was also separated from the Login component and displayed at login when there is no proposal already selected. The user can select a proposal from the menu in the top right corner. 

![Peek 2023-11-22 16-29](https://github.com/mxcube/mxcubeweb/assets/4331447/9c050c94-30e8-4def-a590-fca2d94deaaf)

marcus-oscarsson committed:
Separated login and login_info routes

marcus-oscarsson committed:
Updated tests to use status codes

meguiraun commented:
thanks, just few questions (does not qualify as a proper review)

> The user can select a proposal from the menu in the top right corner.

Is there any situation where the user can play around without selecting a proposal?

And, I am not so sure if we want to change proposals in the middle of the beamtime without a proper logout. (ping @JieNanMAXIV )

fabcor-maxiv reviewed (dismissed):
I guess we will want to properly test this at MAX IV before this gets merged. That is why I put a provisional negative feedback for now, even though I have not reviewed the code yet. And obviously I understand that it is still a draft.

Rework of the authentication code is definitely welcome, it was not straightforward before, thanks for working on this :)

marcus-oscarsson commented:
@meguiraun: I don't think that its possible to proceed without selecting a proposal :). The "change" proposal dialog can of-course be made optional. We will need a feature like this for the automated beamlines where you can collect as several proposals during one session.

@fabcor-maxiv: Sure, understandable take your time and test it. This is the first step there will be a few more updates at a later moment. I'm actually struggling a bit with the session timeout test, while it works in the browser it fails in the test. Maybe you have an idea of what it could be ?

axelboc commented on mxcubeweb/routes/login.py:
```diff
@@ -5,17 +5,13 @@
-    resp = jsonify({"msg": msg})
-    resp.code = 409
-    return resp
+    return make_response(jsonify({"msg": msg}), 200)
```
Maybe there's a reason I'm not seeing, but I'm surprised to see a 200 response code inside a function called `deny_access`. 🤔 Did you mean [403 Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) (which means the user is not "authorized" to access a resource)?

marcus-oscarsson replied:
Yes,  the previous code was not working by the way, doing `resp.code=409` had no effect and the status code was always 200.

I was thinking of using 401, and I tried that but I then realized that we are so far (and maybe we should find a another way to do this), returning an error string to the client with a text message to be shown in the "error box" below the login fields (in the login form).

So, for the moment I kept the 200 and I made it explicit. I, as you, think that we should move away from this and use 401 or 403 instead. But that also means that we need to handle the error display in some other way.

We can perhaps handle that in the PR following this one, what do you think ?

fabcor-maxiv replied:
Bad credentials (can not authenticate) is 401. If user is authenticated (good credentials) but does not have the permission to access resource, then it is 403.

Looking at the context where the `deny_access` function is used, I would argue that `deny_access` is not an appropriate name for this function. It is not the access per se that is denied, it is that the credentials are incorrect. At this point it is not a matter of denying or authorizing access. Maybe `reject_credentials` is a better name.

fabcor-maxiv replied:
Looking at the context where the `deny_access` function is used, I would argue that `deny_access` is not an appropriate name for this function. It is not the access per se that is denied, it is that the credentials are incorrect. At this point it is not a matter of denying or allowing access. Maybe `reject_credentials` is a better name.

fabcor-maxiv replied:
Then regarding the status code...

One can argue that `200` is the correct status code to emit even if the credentials are incorrect. 200 means the http request was handled correctly. Sending incorrect credentials is not an error per se, not an issue with the http request and its handling at least. So sending 200 http code with the body that explicitly tells if the credentials are correct or not `{"msg": "Could not authenticate"}` is fine by me.

axelboc replied:
That's what the HTTP error codes are for, though. 😉 It saves you from sending any message at all, since 401 already implies that the credentials are wrong. It also allows for simpler error handling in the front-end, since the fetching library (wretch in our case, but also axios and others) considers 4xx and 5xx as errors. Not to mention that you get a clear XHR error in the browser dev tools, which helps a lot when debugging.

axelboc replied:
But no worries with using 200 for now and improving error handling later of course! 👍

axelboc replied:
Thought about this a bit more: while responding with 401 for the `/login` endpoint is the way to go (because the purpose of the request is to log in, so if it doesn't work, the server must inform the client with the appropriate error code), it doesn't necessarily make sense to do so for the `/login_info` endpoint.

The `/login_info` endpoint is not an authenticated endpoint; calling it when the user is logged out is valid and expected, so it shouldn't actually respond with 401 but just 200 and a body that conveys that no user is currently logged in (maybe an empty object).

axelboc commented on ui/src/actions/login.js:
```diff
@@ -67,50 +67,39 @@ export function startSession(userInControl) {
+          selectedProposal: '',
+          selectedProposalID: '',
+          loggedIn: false,
+          rootPath: '',
```
This should be an `INITIAL_STATE` constant imported from the reducer file to avoid duplication. Or better, a `resetLoginInfo` action.

marcus-oscarsson replied:
yeah, you are right a `resetLoginInfo` is nicer

axelboc commented on ui/src/actions/login.js:
```diff
@@ -67,50 +67,39 @@ export function startSession(userInControl) {
+    sendLogIn(proposal, password).then(
       (res) => {
-        if (res.code === 'ok') {
+        if (res.msg === '') {
```
This condition should no longer be needed, since 200 (i.e. no error) assumes success and error response codes are handled by `catch` logic below.

marcus-oscarsson replied:
oh yes,

*EDIT:*, ah well ideally yes, but we are actually using the response even if the login was not successful. Ill pass by later and we can see exactly how to deal with this one :)

axelboc replied:
Yep, makes sense now as per your other comment 👍

axelboc commented on ui/src/actions/login.js:
```diff
@@ -125,23 +114,18 @@ export function logIn(proposal, password, navigate) {
-
     return sendSignOut().then(() => {
       dispatch({ type: 'SIGNOUT' });
-      dispatch(getLoginInfo());
```
Do you need to dispatch the action I suggested, `resetLoginInfo`? Or is the `SIGNOUT` action sufficient?

marcus-oscarsson replied:
I think SIGNOUT is indeed sufficient but you are right that something like `resetLoginInfo` should probably called to clean up things.

axelboc commented on ui/src/api/login.js:
```diff
@@ -11,7 +11,12 @@ export function sendSignOut() {
+  return endpoint
+    .get('/login_info') // eslint-disable-next-line promise/prefer-await-to-callbacks
+    .unauthorized((err) => {
+      throw err;
```
Not needed if all you do is re-throw, unless this is perhaps a placeholder to do other things later.

axelboc commented on ui/src/components/App.js:
```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
+  const { applicationFetched } = props;
 
   useEffect(() => {
     requireAuth();
```
Might be cleaner to inline the function in here and use `mapDispatchToProps` to bind the various action creators.

axelboc commented on ui/src/components/App.js:
```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
+    return;
   }
+
+  await store.dispatch(getInitialState());
```
Since the `getInitialState` async action is dispatched only once, when the `App` is rendered for the first time, it would make sense to fetch the login info inside that action as well, in parallel with all the other requests.

You can also move `dispatch(applicationFetched(true));` in there and remove `dispatch(applicationFetched(false));`, I think.

marcus-oscarsson replied:
Hm, that would perhaps be possible but I don't want to fetch the entire app if `fetchLoginInfo` returns 401, or is that just silly of me ?

axelboc replied:
True, true, it's not a bad idea to want to keep them separate, but it does complexify things. If `fetchLoginInfo` fails, then do you dispatch `getInitialState` again when the user logs in? If you don't, I think it's a bug; if you do, then `getInitialState` is perhaps not the right wording.

It's a choice between slowing down initial load, or slowing down logging in. The ideal UX solution is in between: fast initial load, combined with pre-loading while the user enters their crendentials.

In terms of code, fetching everything on initial load is easier since you don't have to worry about dispatching `getInitialState` again. Note that there are probably opportunities for optimising this, for instance, by combining some endpoints to reduce the number of concurrent requests.

axelboc commented on ui/src/components/App.js:
```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
+    store.dispatch(applicationFetched(false));
+    await store.dispatch(getLoginInfo());
+  } catch {
+    store.dispatch(applicationFetched(true));
+    return;
```
If `getLoginInfo` fails, then the app shouldn't remain in a loading state. I think this is how the app behaves currently but it's not great, obviously. Might be good to fix this behaviour as part of this PR.

marcus-oscarsson replied:
I think its solved with the ` store.dispatch(applicationFetched(true));`  application is considered fetched if `getLoginInfo` fails in the code above or when `getInitialState()` is done.

axelboc replied:
Oh right, I must have misread sorry. The error will need to be handled a bit better in a future PR, then. :+1:

axelboc commented on ui/src/components/App.js:
```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
   }
+
+  await store.dispatch(getInitialState());
+  serverIO.listen(store);
```
Need to disconnect in the `useEffect`'s clean-up function to avoid memory leaks: https://react.dev/reference/react/useEffect#parameters

axelboc commented on ui/src/actions/general.js:
```diff
@@ -82,9 +82,12 @@ function notify(error) {
     const state = {};
+    const { login } = getState();
+
+    dispatch(applicationFetched(false));
```
I don't think this is needed, since this is the default value in the store on initialisation, and `getInitialState` is only dispatched once.

axelboc commented on ui/src/reducers/index.js:
```diff
@@ -40,10 +40,6 @@ const mxcubeReducer = combineReducers({
 });
 
 const rootReducer = (state, action) => {
```
This wrapper function is no longer needed I think, you can inline the line below directly.

axelboc commented on ui/src/components/App.js:
```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
+import { serverIO } from '../serverIO';
 
-  if (login.loggedIn) {
-    await store.dispatch(startSession(login.user.inControl));
```
I believe you can now scrap the `startSession` action creator entirely.

marcus-oscarsson replied:
Oh yes, I wanted to do that, thanks for the reminder :)

fabcor-maxiv replied:
Seems like it is done.

marcus-oscarsson committed:
disconnecting socketio in useEffect cleanup

marcus-oscarsson committed:
Added a resetLoginInfo action

marcus-oscarsson committed:
Removed stray print

marcus-oscarsson committed:
Removed deny_access method

marcus-oscarsson committed:
removed unnecessary unauthorized handler

marcus-oscarsson committed:
Added missing session.permanent = True

marcus-oscarsson committed:
Added missing file

marcus-oscarsson commented on mxcubeweb/routes/login.py:
```diff
@@ -45,15 +32,17 @@ def login():
+        else:
+            res = make_response(jsonify({"msg": ""}), 200)
 
         session.permanent = True
```
I managed to get the tests working by adding back this line which I initially removed, as I thought that setting `SESSION_PERMANENT=True` would be enough and certainly as we are also using `SESSION_PERMANENT_LIFETIME` (and the later is taken into account). 

However it seems that the configuration value is not taken into account and we need to set it programmatically, have a look at our mockup server [config](https://github.com/mxcube/mxcubeweb/blob/b38bb1446733ced2ff3fc982aa6f386ee412b47d/test/HardwareObjectsMockup.xml/mxcube-web/server.yaml#L6)

Documentation even states that its true by default: https://flask-session.readthedocs.io/en/latest/config.html#configuration

Does someone have an idea of what that could be ?

fabcor-maxiv replied:
@marcus-oscarsson

I recall it not being straightforward to understand. It took me a while to wrap my head around sessions and their expiration. I had written [some doc](https://mxcubeweb.readthedocs.io/en/latest/dev/login.html), but obviously this doc is not good enough since reading it back now does not clarify much...

Flask and Flask-session have different behavior, different defaults. Flask does not seem to know about the `SESSION_PERMANENT` config setting at all.

I made a quick search and I am starting to wonder if flask-session is used at all. I can not seem to find where it is used in the code. Am I missing something?

marcus-oscarsson replied:
Indeed that explains it, so looking at the [flask documentation](https://flask.palletsprojects.com/en/2.0.x/config/) `SESSION_PARMANENT` is not an option at all, however `SESSION_PERMANENT_LIFETIME` is and its used when one does `session.permanent=True`.  So I was simply confused by the Flask-session documentation, we were using it in the past but now we are using `Flask-Login`

fabcor-maxiv replied:
We should probably remove line 29 from `mxcubeweb/server.py`, line 6 from `test/HardwareObjectsMockup.xml/mxcube-web/server.yaml`, and finally line 20 from `mxcubeweb/core/models/configmodels.py`.

marcus-oscarsson replied:
:), Actually I thought if flask does not deal with the `SESSION_PERMANENT` setting we can :)

marcus-oscarsson replied:
On a second thought, thats just unnecessary ... Ill remove the config option as we are actually relying on this feature, it does not make any sense to set it to false in our case.

fabcor-maxiv replied:
I wonder now if it makes sense (or is needed) to declare the session permanent when authentication failed.

It seems to work as it is so maybe it is fine to leave it like this, but curiosity...

marcus-oscarsson committed:
Removed action creator, startSession

marcus-oscarsson committed:
Removed unused code

marcus-oscarsson committed:
Improved HTTP status code handling, using 401 on all unauthorized resrouces

marcus-oscarsson committed:
Moved getInitialState and related logic to login
Removed circular dependency on serverIO

marcus-oscarsson commented:
Just a few last improvements done after some discussions we had between us here, after this I think we are more or less happy with this one. There will as I mentioned before be one or two more PR's to follow up on this.

 - Status codes of all unauthorized resources are now 401
 - `login_info` always have status code 200 but returns `loggedIn=False|True`
 - Moved `getLoginInfo` call to `getInitialState` and removed unnecessary `navigate`
 - Moved `getInitialState` to `actions/login` to remove a circular dependency cycle (And it it also makes sense to have getIntialState in `reducers/login`
 - Moved the use of `serverIO.disconnect` from `action/login` so that it only contains redux related code and as a consequence avoids a circular dependency

marcus-oscarsson committed:
Fixed doc-string

axelboc commented on ui/src/actions/login.js:
```diff
@@ -124,25 +121,188 @@ export function logIn(proposal, password, navigate) {
+      dispatch(resetLoginInfo());
       if (navigate) {
-        navigate('/login');
+        navigate('/');
```
Would be good to navigate outside of Redux here as well if possible (perhaps for another PR).

axelboc commented on ui/src/actions/login.js:
```diff
@@ -59,58 +83,31 @@ export function sendSelectProposal(number, navigate) {
+  return async (dispatch) => {
+    try {
+      const loginInfo = await fetchLoginInfo();
+      if (!loginInfo.loggedIn) {
+        throw new Error('Not authenticated');
+      }
```
As discussed, I think we can remove this `throw` and the `catch` below, since it cannot happen if the app is coded properly. If we get to this and the user is still unauthenticated, then we will hit a 401 when fetching other things later as expected (e.g. in `getInitialState`).

marcus-oscarsson replied:
:+1:

marcus-oscarsson committed:
Removed unnecessary navigate and removed unused error handling code

axelboc committed:
Review when to get `getLoginInfo` and `getInitialState`

axelboc commented:
I've refactored things a bit to make it clearer when/where `getLoginInfo` and `getInitialState` get called. Here's the gist of it:

- `getLoginInfo` is called from `useEffect` in `App` and from the `logIn` action, which means it gets called:
  1. once on initial load, to know if the user is logged in or not;
  2. right after a successful login (until such times as the `login` endpoint returns the login info right away).
- `getInitialState` is called only from `componentDidMount` in `Main`, which means that it gets called:
  1. on initial load when the user is already logged in;
  2. after a successful login.

In other words, `getInitialState` no longer needs to call `getLoginInfo` directly, since by the time `Main` renders, the user's login info have already been fetched (either from `useEffect` or from `login`). This means we can move `getInitialState` out of the `login.js` actions file without reintroducing a circular import. I'll do it in a separate commit later this week.

fabcor-maxiv commented:
Would that make sense to add some of these technical details at `docs/source/dev/login.rst`?

fabcor-maxiv commented:
For info, I have not been able to work on this these past weeks, but I am planning to test these changes to authentication this week (possibly tomorrow already).

marcus-oscarsson commented:
@fabcor-maxiv what I suggest is that we merge this as it is now and we fix anything we/you find while testing. Unless you see a fundamental problem with this PR

fabcor-maxiv commented:
~I am investigating a case where I get an error `TypeError: t.loginInfo.user is undefined`, seemingly in the `SET_LOGIN_INFO` action. It appears right when loading MXCuBE the first time, before any authentication. It stays stuck on the "Loading, please wait" page. I will provide more details, once I manage to narrow this down further.~

False alarm.

fabcor-maxiv commented:
I figured out the cause of the behavior I initially reported in the previous message, and it was a "user issue" :D

Now it seems to be working fine, comparable to what is demonstrated by the animation in the first post. Nice! :)

fabcor-maxiv commented on mxcubeweb/core/components/user/usermanager.py:
```diff
@@ -275,8 +258,10 @@ def login_info(self):
 
             res["selectedProposalID"] = HWR.beamline.session.proposal_id
+        else:
+            raise Exception("Not logged in")
```
Maybe raise a more specific exception type? See `mxcubeweb/routes/login.py` line 85 where is is caught.

```suggestion
            raise NotAuthenticated("Not logged in")
```

fabcor-maxiv commented on mxcubeweb/routes/login.py:
```diff
@@ -80,24 +68,22 @@ def login_info():
+        try:
+            res = app.usermanager.login_info()
             response = jsonify(res)
+        except Exception:
```
What about having a more specific Exception type? See `mxcubeweb/core/components/user/usermanager.py` line 262 where the exception is raised.

```suggestion
        except NotAuthenticated:
```

fabcor-maxiv commented on test/test_authn.py:
```diff
@@ -77,8 +77,6 @@ def client(make_client):
     resp = client.post(URL_SIGNIN, json=CREDENTIALS_0)
     assert resp.status_code == 200
-    assert resp.json["code"] == "ok"
-    assert resp.json["msg"] == "Successful login"
```
We could put the test of the message back in:

```
    assert resp.json["msg"] == "Successful login"
```

If we do not want the old message then at least we should test that the message is the empty string (as it is now in this pull request):

```
assert resp.json["msg"] == ""
```

fabcor-maxiv commented on mxcubeweb/routes/login.py:
```diff
@@ -66,8 +55,7 @@ def signout():
         app.usermanager.signout()
-
-        return redirect("/login", code=302)
+        return make_response(jsonify(""), 200)
```
Maybe add a meaningful message instead of empty string:

```suggestion
        return make_response(jsonify("Successful sign out"), 200)
```

fabcor-maxiv replied:
OK, after response from @axelboc I understand we could leave it as empty for now, until we get to the point of being able to return the actual login info instead.

fabcor-maxiv commented on mxcubeweb/routes/login.py:
```diff
@@ -45,15 +32,17 @@ def login():
-            res = deny_access("Could not authenticate")
+            res = make_response(jsonify({"msg": "Could not authenticate"}), 200)
+        else:
+            res = make_response(jsonify({"msg": ""}), 200)
```
Why no message?

We could put the old message back in:

```suggestion
            res = make_response(jsonify({"msg": "Successful login"}), 200)
```

axelboc replied:
FYI, if I'm not mistaken, the front-end has an equality check for the empty string, so it will need to be updated as well. This is temporary, though. The idea is to return the login info object directly eventually.

fabcor-maxiv replied:
I see, thanks! Then let's leave it like this for now, we only need to adjust the test.

fabcor-maxiv commented on test/test_authn.py:
```diff
@@ -90,11 +88,9 @@ def test_authn_signin_wrong_credentials(client):
     resp = client.get(URL_SIGNOUT)
-    assert resp.status_code == 302
-    assert resp.headers["Location"] == "/login"
+    assert resp.status_code == 200
```
Let's check the message, no?

```suggestion
    assert resp.status_code == 200
    assert resp.json["msg"] == ""
```

~Even better if we had a meaningful message instead of empty string~:

```suggestion
    assert resp.status_code == 200
    assert resp.json["msg"] == "Successful sign out"
```

fabcor-maxiv reviewed (commented):
Some comments for the Python part

fabcor-maxiv commented on ui/src/actions/login.js:
```diff
@@ -59,90 +83,205 @@ export function sendSelectProposal(number, navigate) {
-      console.log(serverIO.hwrSocket.connected); // eslint-disable-line no-console
-    } else {
-      serverIO.connect();
+    if (res.msg !== '') {
```
Ah, maybe that is the reason why all the success messages were removed on the back end side.

I am not convinced it is a great choice, but I can understand...

Empty message or not, the back end tests should test the message, because as we see here it is obviously part of the API.

axelboc replied:
Ha, should have scrolled down further... Yeah, so the idea will be to return the login info object directly in a future PR.

fabcor-maxiv replied:
Ah, I see. Then @marcus-oscarsson I recommend to adjust the Python tests to assert that we do have an empty string where we expect one (as mentioned in earlier comments).

fabcor-maxiv reviewed (commented):
Comment regarding the empty message seen from the front end side.

fabcor-maxiv commented on test/test_authn.py:
```diff
@@ -77,8 +77,6 @@ def client(make_client):
 def test_authn_signin_good_credentials(client):
     resp = client.post(URL_SIGNIN, json=CREDENTIALS_0)
     assert resp.status_code == 200
```
```suggestion
    assert resp.status_code == 200
    assert resp.json["msg"] == ""
```

marcus-oscarsson committed:
Update test/test_authn.py

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

marcus-oscarsson committed:
Update test/test_authn.py

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

marcus-oscarsson committed:
Fixed test

marcus-oscarsson committed:
Moved session.permanent to successful login

marcus-oscarsson commented:
Ok, thanks guys (@axelboc and @fabcor-maxiv). The work on some of the details will continue in followup-PRs, and fairly soon one that also introduces OpenID Connect  authentication using KeyCloak.

fabcor-maxiv commented:
@axelboc @marcus-oscarsson Seems like we lost the loading animation on the "Sign in" button in this change. I can open a separate ticket if you prefer.

P.S.: ticket opened: https://github.com/mxcube/mxcubeweb/issues/1148
